ncd-calculator/src/services/kgrid.ts

```diff
@@ -0,0 +1,595 @@
+    // 2. Secondary pattern: subtle sinusoidal variation to break additional symmetries
+
+    // Base component with carefully calibrated parameters
+    const baseGradient = 1.15 + 0.25 * normalizedI + 0.35 * normalizedJ;
+
+    // Subtle secondary pattern to break remaining symmetries
+    const secondaryPattern = 0.03 * Math.sin(normalizedI * 3.14) * Math.sin(normalizedJ * 2.71);
+
+    // Combine components with carefully chosen exponent (0.4)
+    // - Higher than 0.2 (your original) for more differentiation
+    // - Lower than 0.6 to avoid overwhelming the NCD relationships
+    return Math.pow(baseGradient + secondaryPattern, 0.4);
+};
+
+export const calculateObjective = (grid: GridState): number => {
+    let total = 0;
+
+    for (let i = 0; i < grid.height; i++) {
+        for (let j = 0; j < grid.width; j++) {
+            const currentId = grid.grid[i][j];
+            let positionNCD = 0;
+
+            const rightJ = (j + 1) % grid.width;
+            const rightNeighborId = grid.grid[i][rightJ];
+
+            positionNCD += grid.ncdMatrix[currentId][rightNeighborId];
+
+            const downI = (i + 1) % grid.height;
+            const downNeighborId = grid.grid[downI][j];
+            positionNCD += grid.ncdMatrix[currentId][downNeighborId];
+
+            const factor = gradualFactor({i, j}, grid.width, grid.height);
+            total += positionNCD * factor;
+
+        }
+    }
+    return total;
+}
+
+export const deepCopy = (gridState: GridState): GridState => {
+    const newGrid: string[][] = [];
+
+    for (let i = 0; i < gridState.height; i++) {
+        newGrid[i] = [];
+        for (let j = 0; j < gridState.width; j++) {
+            newGrid[i][j] = String(gridState.grid[i][j]);
+        }
+    }
+    return {
+        width: gridState.width,
+        height: gridState.height,
+        ncdMatrix: gridState.ncdMatrix,
+        grid: newGrid,
+        objectiveValue: gridState.objectiveValue
+    }
+}
+
+export const extractBlock = (grid: GridState, block: Block): string[][] => {
+    // Normalize the block to ensure proper coordinates
+    const normalizedBlock = normalizeBlock(block, grid.height, grid.width);
+    const {topLeft, bottomRight} = normalizedBlock;
+
+    // Calculate exact dimensions
+    const blockHeight = bottomRight.i - topLeft.i + 1;
+    const blockWidth = bottomRight.j - topLeft.j + 1;
+
+    console.log(`Extracting block with dimensions ${blockWidth}×${blockHeight}`);
+
+    // Create a properly sized array
+    const blockContent: string[][] = Array(blockHeight).fill(null)
+        .map(() => Array(blockWidth).fill(""));
+
+    // Fill the array with grid content
+    for (let i = 0; i < blockHeight; i++) {
+        for (let j = 0; j < blockWidth; j++) {
+            // Apply toroidal wraparound properly
+            const gridI = (topLeft.i + i) % grid.height;
+            const gridJ = (topLeft.j + j) % grid.width;
+
+            blockContent[i][j] = String(grid.grid[gridI][gridJ]);
```

Comment:
i think this string should be integer.  the block extract doesn't need to know contents. it just needs integer indexes so it can look up ncd values for cost calculation later.